ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh

```diff
@@ -4,10 +4,376 @@ set -o nounset
+        ;;
+esac
+
 function run_command() {
     local CMD="$1"
     echo "Running Command: ${CMD}"
-    eval "${CMD}"
+    # Execute command and capture both output and exit code
+    local output
+    output=$(eval "${CMD}" 2>&1)
+    local exit_code=$?
+    
+    if [[ ${exit_code} -ne 0 ]]; then
+        echo "WARNING: Command failed: ${CMD} (exit code: ${exit_code})"
+        echo "Output: ${output}"
+        return ${exit_code}
+    fi
+    
+    # Return the output for command substitution
+    echo "${output}"
+}
+
+function run_command_silent() {
+    local CMD="$1"
+    # Execute command silently and return output for command substitution
+    bash -c "${CMD}" 2>/dev/null || echo ""
+}
+
+function wait_for_condition() {
+    local condition_name="$1"
+    local check_command="$2"
+    local max_attempts="$3"
+    local sleep_interval="$4"
+    local failure_message="$5"
+    local show_status_command="$6"  # Optional command to show status (both during waiting and on success)
+    
+    local attempt=0
+    
+    log "Waiting for ${condition_name}..."
+    
+    while [[ ${attempt} -lt ${max_attempts} ]]; do
+        if eval "${check_command}"; then
+            log "✅ ${condition_name} condition met"
+            
+            # Show success details if command provided
+            if [[ -n "${show_status_command}" ]]; then
+                log ""
+                log "📊 ${condition_name} success details:"
+                eval "${show_status_command}" || true
+                log ""
+            fi
+            
+            return 0
+        fi
+        
+        log "⏳ ${condition_name} not ready, waiting... (${attempt}/${max_attempts})"
+        
+        # Show current status if command provided and at specific intervals
+        if [[ -n "${show_status_command}" ]] && [[ ${attempt} -eq 0 ]] || [[ $((attempt % 5)) -eq 0 ]]; then
+            log ""
+            log "📋 Current ${condition_name} status:"
+            eval "${show_status_command}" || true
+            log ""
+        fi
+        
+        sleep "${sleep_interval}"
+        attempt=$(( attempt + 1 ))
+    done
+    
+    log "❌ ERROR: ${failure_message}"
+    return 1
+}
+
+
+function get_infra_id() {
+    local infra_id=""
+    if [[ -f "${SHARED_DIR}/metadata.json" ]]; then
+        infra_id=$(jq -r '.infraID' "${SHARED_DIR}/metadata.json")
+    fi
+    echo "${infra_id}"
+}
+
+# Function to wait for cluster to be fully recovered after reboot
+function wait_for_cluster_recovery() {
+    log "=========================================="
+    log "Waiting for cluster to fully recover after reboot"
+    log "=========================================="
+    
+    # Step 1: Wait for API server to be accessible
+    log "Step 1: Waiting for API server to be accessible..."
+    if ! wait_for_condition "API server connectivity" \
+        "oc whoami >/dev/null 2>&1" \
+        "${CLUSTER_RECOVERY_TIMEOUT}" 30 \
+        "API server not accessible within 60 minutes" \
```

Comment:
If the API is not accessible, the next oc command or the next health check step would fail. So an individual API-check step is not necessary